atlasdb-dbkvs-hikari/src/main/java/com/palantir/nexus/db/pool/HikariCPConnectionManager.java

```diff
@@ -244,12 +246,19 @@ private void testDataSource(HikariDataSource dataSourcePool) throws SQLException
      * @throws SQLException if connection is invalid and cannot be used.
      */
     private void testConnection(Connection connection) throws SQLException {
-        try (Statement stmt = connection.createStatement();
-                ResultSet rs = stmt.executeQuery(connConfig.getTestQuery())) {
-            if (!rs.next()) {
-                throw new SQLException(String.format(
-                        "Connection %s could not be validated as it did not return any results for test query %s",
-                        connection, connConfig.getTestQuery()));
+        if (connConfig.getTestQuery().isEmpty()) {
+            if (!connection.isValid(connConfig.getSocketTimeoutSeconds())) {
+                throw new SQLException(String.format("Connection %s is not valid", connection));
+            }
+        } else {
+            try (PreparedStatement stmt = connection.prepareStatement(connConfig.getTestQuery())) {
+                stmt.setFetchSize(1);
```

Comment:
Probably worth some test runs for Oracle and Postgres, but https://blog.jooq.org/setting-the-jdbc-statement-setfetchsize-to-1-for-single-row-queries/ makes the case that 1 is actually too low for Oracle, though we may want to just skip setting fetch size and use driver default.

```suggestion
                stmt.setFetchSize(2);
```

Reply:
I wasn't able to replicate the results in that article (using the latest Oracle JDBC driver, 19.15 release).

Setting a fetch size doesn't seem to make any material difference, nor is isValid any faster (although this might be because the test is a bit fake since I'm just looping each test 10k times).

At the end of the day, I'm much more interested in getting rid of the checkout test at all, because that is where there is a real gain to be made.

SELECT 1 default fetch 00:00:14.127
SELECT 1 fetch 1 00:00:13.756
SELECT 1 fetch 2 00:00:15.779
SELECT x default fetch 00:00:15.430
isValid 00:00:15.009

SELECT 1 default fetch 00:00:17.406
SELECT 1 fetch 1 00:00:18.338
SELECT 1 fetch 2 00:00:18.500
SELECT x default fetch 00:00:17.416
isValid 00:00:15.633

SELECT 1 default fetch 00:00:14.398
SELECT 1 fetch 1 00:00:16.419
SELECT 1 fetch 2 00:00:17.542
SELECT x default fetch 00:00:18.172
isValid 00:00:16.740